server/service/src/document/encounter/insert.rs

```diff
@@ -0,0 +1,275 @@
+    pub data: serde_json::Value,
+    pub schema_id: String,
+}
+
+pub fn insert_encounter(
+    ctx: &ServiceContext,
+    service_provider: &ServiceProvider,
+    store_id: String,
+    user_id: &str,
+    input: InsertEncounter,
+) -> Result<Document, InsertEncounterError> {
+    let patient = ctx
+        .connection
+        .transaction_sync(|_| {
+            validate(ctx, &store_id, &input)?;
+            let doc = generate(user_id, input)?;
+
+            // Updating the document will trigger an update in the patient (names) table
+            let result = service_provider
+                .document_service
+                .update_document(ctx, &store_id, doc)
+                .map_err(|err| match err {
+                    DocumentInsertError::InvalidDataSchema(err) => {
+                        InsertEncounterError::InvalidDataSchema(err)
+                    }
+                    DocumentInsertError::DatabaseError(err) => {
+                        InsertEncounterError::DatabaseError(err)
+                    }
+                    DocumentInsertError::InternalError(err) => {
+                        InsertEncounterError::InternalError(err)
+                    }
+                    _ => InsertEncounterError::InternalError(format!("{:?}", err)),
+                })?;
+
+            Ok(result)
+        })
+        .map_err(|err: TransactionError<InsertEncounterError>| err.to_inner_error())?;
+    Ok(patient)
+}
+
+impl From<RepositoryError> for InsertEncounterError {
+    fn from(err: RepositoryError) -> Self {
+        InsertEncounterError::DatabaseError(err)
+    }
+}
+
+fn generate(user_id: &str, input: InsertEncounter) -> Result<RawDocument, RepositoryError> {
+    let encounter_name = Utc::now().to_rfc3339();
+    Ok(RawDocument {
+        name: patient_program_encounter_doc_name(&input.patient_id, &input.r#type, &encounter_name),
+        parents: vec![],
+        author: user_id.to_string(),
+        timestamp: Utc::now(),
+        r#type: input.r#type,
+        data: input.data,
+        schema_id: Some(input.schema_id),
+    })
+}
+
+fn validate_encounter_schema(
+    input: &InsertEncounter,
+) -> Result<SchemaEncounter, serde_json::Error> {
+    serde_json::from_value(input.data.clone())
+}
+
+fn validate_patient_program_exists(
+    ctx: &ServiceContext,
+    store_id: &str,
+    patient_id: &str,
+    program: &str,
+) -> Result<bool, RepositoryError> {
+    let doc_name = patient_program_doc_name(patient_id, program);
+    let document =
+        DocumentRepository::new(&ctx.connection).find_one_by_name(store_id, &doc_name)?;
+    Ok(document.is_some())
+}
+
+fn validate(
+    ctx: &ServiceContext,
+    store_id: &str,
+    input: &InsertEncounter,
+) -> Result<(), InsertEncounterError> {
+    if !validate_patient_program_exists(ctx, store_id, &input.patient_id, &input.program_type)? {
+        return Err(InsertEncounterError::InvalidPatientOrProgram);
+    }
+
+    validate_encounter_schema(input).map_err(|err| {
```

Comment:
I guess this will be validating against actual encounter schema eventually ?

Reply:
It actually does it in the document service. I added a comment.

Reply:
But validate_encounter_schema uses `SchemaEncounter` type from `./schema/ecounter`, sorry I don't quite understand I thought documents were configurable.

Reply:
So the validation that derive encounter schema is compatible with base encounter schema is done at the time of encounter insertion ? Shouldn't we check that at the time of configuration (and on insert/update validation is done against derived schema ?)  ? As per comment above it seems like a database structures is the best way to enforce required data like this.

Reply:
Having it checked at time of configuration the schema would be ideal. However, haven't found an easy way to do that correctly yet but I haven't looked into it too much.